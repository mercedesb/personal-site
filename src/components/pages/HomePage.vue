<template>
  <transition name="fade">
  <div class="u-fillHeight">
    <HeroHeader
      :title="page.preamble"
      color='brown'
      :mainImage="mainImageUrl"
      :backgroundImages="backgroundImages"></HeroHeader>
    <div class='FlexContainer FlexContainer--column u-fillHeight'>
    <div class='FlexContainer u-fillSpace' v-if="columns && columns.length">
        <ContentColumn
          v-for="column in columns"
          :key="column.id"
          v-bind="column"
        />
    </div>
  </div>
</div>
</transition>
</template>

<script>
import HeroHeader from '../HeroHeader.vue'
import ContentColumn from '../ContentColumn.vue'

export default {
  components: {
    HeroHeader, ContentColumn
  },
  computed: {
    page() {
      return this.$store.state.homePage
    },
    mainImageUrl() {
      return this.getImageUrl(this.page.mainImage)
    },
    backgroundImages() {
      if (!this.page.backgroundImages) return []
      return this.page.backgroundImages.map((bgdImage) => {
          return this.getImageUrl(bgdImage)
        })
    },
    columns() {
      if (!this.page.children) return []
      return this.page.children.map((child) => {
        return {
          id: child.sys.id,
          title: child.fields.title,
          color: child.fields.color,
          preamble: child.fields.preamble,
          icon: child.fields.icon,
          externalLink: child.fields.externalLink,
          urlSegment: child.fields.urlSegment
        }
      })
    }
  },
  created () {
    this.$store.dispatch('getHomePage')
  }
}
</script>

<style lang="scss">
</style>