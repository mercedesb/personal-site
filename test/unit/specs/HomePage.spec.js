import HomePage from '@/components/pages/HomePage.vue'
import { TestUtility } from './TestUtility'

describe('HomePage', () => {
  let component

  const state = {
    homePage: {
      id: 'homepage',
      preamble: 'here is the dev homepage preamble, it\'s a little long to be more realistic',
      mainImage: {
        fields: {
          file: {
            url: 'http://placeholder.pics/svg/300'
          }
        }
      },
      children: Object.keys(TestUtility.landingPages).map((key, index) => {
        return {
          sys: {
            id: `child ${index}`,
          },
          fields: {
            ...TestUtility.landingPages[key]
          }
        }
      }),
      backgroundImages: [
        {
          fields: {
            file: {
              url: 'http://placeholder.pics/svg/300'
            }
          }
        },
        {
          fields: {
            file: {
              url: 'http://placeholder.pics/svg/300'
            }
          }
        },
        {
          fields: {
            file: {
              url: 'http://placeholder.pics/svg/300'
            }
          }
        }
      ]
    }
  }

  const actions = {
    getHomePage: jest.fn()
  }

  const store = {
    state,
    actions
  }

  const shallow = propsData => TestUtility.shallow(HomePage, {
    store,
    propsData: {
      ...propsData
    }
  }) 

  describe('Snapshots', () => {
    it('matches snapshot', () => {
      component = shallow()
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('Computed', () => {
    describe('page', () => {
      beforeEach(() => {
        component = shallow()
      })

      it('has page populated', () => {
        expect(component.vm.page).toBeDefined()
      })
    })

    describe('mainImageUrl', () => {
      beforeEach(() => {
        component = shallow()
      })

      it('has mainImageUrl populated', () => {
        expect(component.vm.mainImageUrl).toBeDefined()
      })
    })

    describe('backgroundImages', () => {
      describe('with background images in cms', () => {
        it('has backgroundImages populated', () => {
          component = shallow()
          expect(component.vm.backgroundImages).toBeDefined()
        })
      })

      describe('without background images in cms', () => {
        it('has backgroundImages populated', () => {
          const localHomePage = {
            ...state.homePage,
            backgroundImages: undefined
          }
          const localStore = {
            state: {
              homePage: localHomePage,
            },
            actions
          }
          component = TestUtility.shallow(HomePage, {
            store: localStore
          }) 

          expect(component.vm.backgroundImages).toBeDefined()
        })
      })
    })

    describe('columns', () => {
      describe('with children in cms', () => {
        it('has columns populated', () => {
          component = shallow()
          expect(component.vm.columns).toBeDefined()
        })

        it('creates objects of the desired shape', () => {
          component = shallow()
          const expectedShape = {
            id: expect.any(String),
            title: expect.any(String),
            color: expect.any(String),
            preamble: expect.any(String),
            icon: expect.any(Object),
            externalLink: expect.any(String),
            urlSegment: expect.any(String)
          }

          expect(component.vm.columns[0]).toMatchObject(expectedShape)
        })
      })

      describe('without children in cms', () => {
        it('has columns populated', () => {
          const localHomePage = {
            ...state.homePage,
            children: undefined
          }
          const localStore = {
            state: {
              homePage: localHomePage,
            },
            actions
          }
          component = TestUtility.shallow(HomePage, {
            store: localStore
          }) 

          expect(component.vm.columns).toBeDefined()
        })
      })
    })
  })

  describe('Lifecycle', () => {
    describe('created', () => {
      it('dispatches getHomePage to the store', () => {
        component = TestUtility.mount(HomePage, { 
          store
         })
        expect(component.vm.$store.dispatch).toHaveBeenCalledWith('getHomePage')
      })
    })
  })
})