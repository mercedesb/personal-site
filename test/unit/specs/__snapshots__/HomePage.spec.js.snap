// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Snapshots matches snapshot 1`] = `
<div
  class="Home"
>
  <!---->
   
  <div
    class="FlexContainer FlexContainer--column u-fillHeight"
  >
    <div
      class="FlexContainer u-fillSpace"
    >
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </div>
</div>
`;
